pkg/serverless/protocol.go

```diff
@@ -111,49 +113,54 @@ type LogsCollection struct {
 
 // ServeHTTP - see type LogsCollection comment.
 func (l *LogsCollection) ServeHTTP(w http.ResponseWriter, r *http.Request) {
+	// If the DogStatsD daemon isn't ready, wait for it.
+	l.daemon.ReadyWg.Wait()
+
 	data, _ := ioutil.ReadAll(r.Body)
 	defer r.Body.Close()
 	var messages []aws.LogMessage
+
 	if err := json.Unmarshal(data, &messages); err != nil {
 		log.Error("Can't read log message")
 		w.WriteHeader(400)
 	} else {
 		metricsChan := l.daemon.aggregator.GetBufferedMetricsWithTsChannel()
-		functionARN := aws.GetARN()
-		functionName := aws.FunctionNameFromARN()
-		// FIXME(remy): could this be exported to properly get all tags?
-		metricTags := []string{
-			fmt.Sprintf("functionname:%s", functionName),
-			fmt.Sprintf("function_arn:%s", functionARN),
-		}
+		metricTags := getTagsForEnhancedMetrics()
+		sendLogsToIntake := config.Datadog.GetBool("logs_enabled")
 		for _, message := range messages {
-			switch message.Type {
-			case aws.LogTypePlatformStart:
+			// Do not send logs or metrics if we can't associate them with an ARN or Request ID
+			// First, if the log has a Request ID, set the global Request ID variable
+			if message.Type == aws.LogTypePlatformStart {
 				if len(message.ObjectRecord.RequestID) > 0 {
 					aws.SetRequestID(message.ObjectRecord.RequestID)
 				}
-				l.ch <- message
+			}
+			// If the global request ID or ARN variable isn't set at this point, do not process further
+			if aws.GetARN() == "" || aws.GetRequestID() == "" {
+				continue
```

Comment:
I looked into 500'ing here. The advantage is that if we 500 the Lambda Logs API will retry sending the logs later. However, things get complicated. Logs arrive in batches. A batch of logs could contain a `START` log line which we need to set the Request ID. So in this scenario:

```
random log line
START log
```

If we 500 when we see the random log line, we will never get the START log line.

In practice, I think we are inevitably going to drop logs before the first function invocation begins (i.e. logs produced when the environment starts for the first time during a cold start) because we have nothing to associate them with. But then we should be good from that first invocation onwards.